### PR TITLE
Feature/issue 15 user opinions exposed via json

### LINKED ISSUE #15: User Opinions are stored and exposed via API
This should be one row in a model with 6 columns (5 opinions and one updated_at column). Pre-populate with "".

```
{"opinions": 
    {
        "1": "Opinion one",
        "2": "Opinion two",
        "3": "Opinion three",
        "4": "Opinion four",
        "5": "Opinion five"
    }
}
```

### PR DESCRIPTION
**Pull request by gitumarkk**
Fix #15 
## Done
- Created opinions directory and added it to settings.
- Added Opinion model.
- Did Schemamigrations on it
- Added API path to urls.py
- Created unittests for API
- Coded the API and made sure it works.
## Assumptions
- The API spits out data in the form of `"opinion_1": "This is opinion 1"` as opposed to  `"1": "This is opinion 1"`
- There is always going to be an opinion hence a situation where there is no opinion has not been considered i.e. {"There is no opinion this week"}.

### CODE REVIEW

**Comment by gitumarkk** [comment]
@imsickofmaps Created opinion app and did a pull request.

File: opinions/tests.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+This file demonstrates writing tests using the unittest module. These will pass
+when you run "manage.py test".
+
+Replace this with more appropriate tests for your application.
+"""
+
+from django.test import TestCase
+
+
+class SimpleTest(TestCase):
```
Comment (imsickofmaps) [inline opinions/tests.py]: @gitumarkk This stub can be cleared out

**Comment by imsickofmaps** [comment]
:+1: apart from excess test

File: opinions/admin.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+from opinions.models import (Opinion,
+                             OpinionPollId,
+                             OpinionPollOpinion,
+                             OpinionPollChoices)
+from django import forms
+from django.forms.models import BaseInlineFormSet
+
+
+class OpinionPollIdForms(forms.ModelForm):
+    pass
```
Comment (smn) [inline opinions/admin.py]: If you're not defining any custom behaviour in this ModelForm then perhaps just omit it?

File: opinions/admin.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+
+class OpinionPollOpinionForms(forms.ModelForm):
+    """
+    The form that is used by OpinionPollIdAdmin
+    """
+    pass
```
Comment (smn) [inline opinions/admin.py]: Same applies here.

File: opinions/admin.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                raise forms.ValidationError("You need atleast one opinion")
+            else:
+                if ("choices" in form.cleaned_data):
+                    char_lim_op = char_lim_op + len(form.cleaned_data['choices'])
+                    if char_lim_op > 160:
+                        raise forms.ValidationError("You have gone beyond the"
+                                                    " character limit"
+                                                    " please shorten opinions"
+                                                    " and/or choices")
+            number_of_choices = + 1
```
Comment (smn) [inline opinions/admin.py:36]: Doesn't the [max_length](https://github.com/praekelt/go-nike-ghr/blob/1b2927e6ccff6ceba1104369efe8a27c6861a7de/opinions/models.py#L52) you've specified already take care of all this hoopla?

File: opinions/admin.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        super(OpinionPollChoicesFormset, self).clean()
+        char_lim_op = len(self.instance.opinion)  # opinion char length
+
+        number_of_choices = 0  # Variable to ensure atleast one choice is entered
+        for form in self.forms:
+            if not hasattr(form, 'cleaned_data'):
+                continue
+
+            if ("choices" not in form.cleaned_data and number_of_choices < 1):
+                raise forms.ValidationError("You need atleast one opinion")
```
Comment (smn) [inline opinions/admin.py]: Typo "at least"

File: opinions/api.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        form by removing the extra meta variables and editing the
+        dictionary
+        """
+        if isinstance(data_dict, dict):
+            if 'meta' in data_dict:
+                del(data_dict['meta'])
+
+            data_dict['opinions'] = (copy.copy(data_dict['objects'][0]))
+            del (data_dict['objects'])
+        return data_dict
```
Comment (smn) [inline opinions/api.py:35]: Not sure if you guys have already coded up the JS bits of this but I'd really suggest leaving all this mangling to the JS app side of things. The idea of REST is to expose resources with data, this goes into application / implementation details too quickly.

File: opinions/api.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+        queryset = OpinionPollId.objects.all()
+
+    def get_object_list(self, request):
+        """
+        tastypie function that returns filtered data
+        """
+        query = super(OpinionPollIdResource, self).get_object_list(request)
+        query = (query.filter(active=True))
+        return query
```
Comment (smn) [inline opinions/api.py]: how about `return query.filter(active=True)` ?

File: opinions/api.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        if isinstance(data_dict, dict):
+            if 'meta' in data_dict:
+                del(data_dict['meta'])
+            if data_dict["objects"] == []:
+                data_dict['quiz'] = False
+                del (data_dict['objects'])
+
+            else:
+                opinions = {}
+                for op_id_i in range(len(data_dict['objects'])):
```
Comment (smn) [inline opinions/api.py]: How about:

``` python
for op_view_data in data_dict['objects']:
    opinion_view_id = "opinion_view_%s" % op_view_data.data["id"]
    ...
```

instead?

File: opinions/api.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+                    opinion_view_dict = {}
+                    op_view_data = data_dict['objects'][op_id_i]
+                    opinion_view_id = "opinion_view_%s" % op_view_data.data["id"]
+                    op_opinions = op_view_data.data["opinionpollid"]
+
+                    start_opinion_id = min([op_opinions[i].data["id"]
+                                           for i in range(len(op_opinions))])
+                    exit_opinion_id = max([op_opinions[i].data["id"]
+                                          for i in range(len(op_opinions))])
```
Comment (smn) [inline opinions/api.py]: Please read up a bit on how Python iterators & list comprehensions work. Your use of `range(len(...))` works but it's unnecessarily complex.

I think the following would work just as well:

``` python
opinion_ids = [data["id"] for data in op_opinions]
start_opinion_id = min(opinion_ids)
exit_optinion_id = max(opinion_ids)
```

File: test_opinion_poll.json [hunk 10]
```diff
@@ -0,0 +1 @@
+[
```
Comment (smn) [inline test_opinion_poll.json]: this doesn't look right

Comment (gitumarkk) [inline test_opinion_poll.json]: @smn not sure what happened there as everything seems okay in the branch https://github.com/praekelt/go-nike-ghr/blob/feature/issue-15-user-opinions-exposed-via-json/opinions/fixtures/test/test_opinion_poll.json

File: opinions/admin.py (hunk 4 above)
Comment (gitumarkk) [inline opinions/admin.py:36]: @smn the max_length is for the individual fields, what I am checking is for the aggregate i.e. len(question) + len(answer_1) + len(answer_2) < 160, which was what I assumed needs to be less than 160 (I could be wrong)

Comment (smn) [inline opinions/admin.py:36]: Nevermind, this is the total number of amount of characters for a single USSD screen, hence the adding up of the total amount of chars.

File: opinions/api.py (hunk 6 above)
Comment (smn) [inline opinions/api.py:35]: Leaving this here for the future us.

File: test_opinion_poll.json (hunk 10 above)
Comment (smn) [inline test_opinion_poll.json]: It probably was a stray file in the root of the repository that got committed by accident.

**Comment by smn** [comment]
@gitumarkk please have a look at the changes I've made and let me know if you're happy.

**Pull request merged**